elks/arch/i86/drivers/char/console-direct.c

```diff
@@ -41,34 +49,86 @@
 
 #define MAXPARMS        28
 
+#define N_DEVICETYPES   2
+#define MAX_OUTPUTS     2
+
+struct output;
+typedef struct output Output;
 struct console;
 typedef struct console Console;
 
+enum OutputType {
+    OT_MDA = 0,
+    OT_CGA,
+    OT_EGA,
+    OT_VGA,
+};
+
 struct console {
+    unsigned char op;
     int cx, cy;                 /* cursor position */
     void (*fsm)(Console *, int);
     unsigned char attr;         /* current attribute */
     unsigned char XN;           /* delayed newline on column 80 */
-    unsigned int vseg;          /* video segment for page */
-    int basepage;               /* start of video ram */
+    unsigned int vseg;          /* vram for this console page */
+    int vseg_offset;            /* vram offset of vseg for this console page */
 #ifdef CONFIG_EMUL_ANSI
     int savex, savey;           /* saved cursor position */
     unsigned char *parmptr;     /* ptr to params */
     unsigned char params[MAXPARMS];     /* ANSI params */
 #endif
+    enum OutputType type;
```

Comment:
I hate to say this, but enum's are problematic. Sometimes an enum is a byte, sometimes its a word. This causes confusion. There are no enums used in the kernel at this time. See my [D-Flat project](https://github.com/ghaerr/dflat/blob/master/system.h#L78) for an example of an enum incompatibility mess between compilers.

Reply:
Yeah, the vague size always bothered me too. I could just #define the different types, and do it that way.

Reply:
Is the bitfield feature supported in `ia16-elf-gcc`? Could make it explicitly one byte with `enum outputType type : 8;` in the struct there?

Reply:
> Could make it explicitly one byte with enum outputType type : 8; in the struct there?

Definitely no bitfields, please. You're new to the project, but we don't want to make the system dependent on a compiler feature when used in only a single place in the kernel. There's just no need for it, and bitfields, enums and various other C features just complicate things. (Yes, I know both are basic nowadays, but we are strive for simplicity around here. If something can be done without invoking a special compiler feature, that's just one less problem for the future when the compiler is changed or the contributor absent (yes, the ELKS compiler did change about 4 years ago, and we are using some GCC extensions for the time being but I'm trying to reduce them. We have a side project trying to bring in the OpenWatcom C compiler for applications which is working - next step is kernel)).

Reply:
Totally fair! I'll go with `#define` for this, should work quite nicely.

Reply:
Fixed in 6ba7c8a8